### PR TITLE
[FLINK-6536] [cep] Improve error message in SharedBuffer::put().

### PR DESCRIPTION
**Pull request by kl0u**
There are two commits that address both FLINK-6536 and FLINK-6255.

R @aljoscha

### CODE REVIEW

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java [hunk 1]
```diff
@@ -96,10 +96,11 @@ public void put(
 
 		// sanity check whether we've found the previous element
 		if (previousSharedBufferEntry == null && previousValue != null) {
-			throw new IllegalStateException("Could not find previous shared buffer entry with " +
+			throw new IllegalStateException("Could not find previous entry with " +
 				"key: " + previousKey + ", value: " + previousValue + " and timestamp: " +
-				previousTimestamp + ". This can indicate that the element belonging to the previous " +
-				"relation has been already pruned, even though you expect it to be still there.");
+				previousTimestamp + ". This can indicate that either you did not implement " +
+				"the equals() and hashCode() methods, of your input elements properly or that " +
```
Comment (aljoscha) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: I think the comma is superfluous.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java [hunk 2]
```diff
@@ -96,10 +96,11 @@ public void put(
...
 		// sanity check whether we've found the previous element
 		if (previousSharedBufferEntry == null && previousValue != null) {
-			throw new IllegalStateException("Could not find previous shared buffer entry with " +
+			throw new IllegalStateException("Could not find previous entry with " +
 				"key: " + previousKey + ", value: " + previousValue + " and timestamp: " +
-				previousTimestamp + ". This can indicate that the element belonging to the previous " +
-				"relation has been already pruned, even though you expect it to be still there.");
+				previousTimestamp + ". This can indicate that either you did not implement " +
+				"the equals() and hashCode() methods of your input elements properly or that " +
+				"the element belonging to that entry has been already pruned.");
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: Been thinking about this exception message and those two reasons for the exception come from "two different worlds". 

1. if element has been pruned - from user perspective this means some error in CEP library occured and the user can not do much about it.
2. the equals() and hashCode() are obvious user mistakes.

Not sure if we should do sth with it, but just wanted to express my thoughts on it.

Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: You are right @dawidwys but I think that now the message makes clear that you should always implement `equals()` and `hashCode()`. So far it was only in the docs.

**Comment by kl0u** [comment]
Merged.

**Pull request closed without merging**